Interaction regions can be outside of parent’s when overflow: hidden is set on the parent

mwyrzykowski opened the pull request:
#### 19f261f96bd352a2017a057a30270cdeb36d2c8a
<pre>
Interaction regions can be outside of parent’s when overflow: hidden is set on the parent
<a href="https://bugs.webkit.org/show_bug.cgi?id=243630">https://bugs.webkit.org/show_bug.cgi?id=243630</a>

Reviewed by Simon Fraser.

Clip the event regions to the background rect if the clipToDirtyRect flag is set. This
is similar to how clipToDirtyRect is used elsewhere throughout RenderLayer. Otherwise
the clip rect for the event regions does not match what is rendererd.

* LayoutTests/interaction-region/event-region-overflow-expected.txt: Added.
* LayoutTests/interaction-region/event-region-overflow.html: Added.
Add a test case.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::collectEventRegionForFragments):
Clip to the dirty rect if flag is set, similar to what RenderLayer does elsewhere.

* LayoutTests/fast/scrolling/mac/negative-z-index-overflow-scroll-expected.txt:
* LayoutTests/editing/editable-region/fixed-and-absolute-contenteditable-scrolled-expected.txt:
* LayoutTests/editing/editable-region/relative-inside-fixed-contenteditable-scrolled-expected.txt:
* LayoutTests/editing/editable-region/transformed-scrolled-on-top-of-fixed-contenteditables-expected.txt:
Update event region expectation as it matches the containing bounds, which seems
like a progression.

* LayoutTests/tiled-drawing/scrolling/non-fast-region/wheel-handler-in-columns-expected.txt:
Previous event regions were incorrect based on visual analysis. Update expectations.

Canonical link: <a href="https://commits.webkit.org/253580@main">https://commits.webkit.org/253580@main</a>
</pre>














<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87171573e16fda5d07e121575ec9bf0a9ed19818

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30358 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95286 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/148994 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90426 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28758 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78592 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90529 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92054 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23390 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66394 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26671 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26585 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13598 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2550 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28264 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28204 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32881 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -86,7 +86,7 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         element = linkElement;
 
-    if (!element || !element->renderer())
+    if (!element || !element->renderer() || element->beforePseudoElement())
```
Does it make sense to ignore all :before? Can they not be important interactible things? Or is it something about the style of it? If yes for :before, why not :after? Might be good to have someone who knows how people build web pages chime in.

mwyrzykowski replied:
Oh I'm not sure. I can tighten the check with `if (!element || !element->renderer() || (element->beforePseudoElement() && !element->countChildNodes()))`, not sure if that would represent a region we are interested in computing

mwyrzykowski replied:
Its certainly possible to construct cases where ignoring ::before is wrong.

I'm taking another look to see what else I can do

mwyrzykowski replied:
@hortont424 clipping the rect to the paint rect resolves the issue I am seeing too but not sure if that is correct.

@smfr @anttijk any thoughts on this?

smfr reviewed (commented):
There are tests for event regions. You should write a test case.

smfr commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -110,10 +112,14 @@ bool EventRegionContext::contains(const IntRect& rect) const
-        if (auto interactionRegion = interactionRegionForRenderedRegion(renderer, region)) {
+        Region clippedRegion = region;
+        if (paintInfo)
+            clippedRegion.intersect(approximateAsRegion(renderer.style().getRoundedBorderFor(paintInfo->rect)));
```
`renderer.style().getRoundedBorderFor(paintInfo->rect)` is certainly wrong; this will construct a rounded-cornered rect using the radii in the style for some arbitrary rect which might be a subset of element's rect.

The other question here is whether intersecting with `paintInfo->rect` is valid. When doing the event regions paint pass, do we always start with a layer-sized clipRect, or do we re-use the same, possibly-smaller clip rect that is used for painting, if this is an incremental paint?

smfr replied:
Also if intersecting with `paintInfo->rect` is correct, I'd expect it to happen out in Renderer code, not down here in EventRegion. And why didn't the event regions paint pass already apply this clip?

mwyrzykowski replied:
Maybe that is the real bug: the event regions paint pass is not applying the clip in this case

mwyrzykowski committed:
Interaction regions can be outside of parent’s when overflow: hidden is set on the parent
https://bugs.webkit.org/show_bug.cgi?id=243630

Reviewed by Simon Fraser.

Clip the event regions to the background rect if the clipToDirtyRect flag is set. This
is similar to how clipToDirtyRect is used elsewhere throughout RenderLayer. Otherwise
the clip rect for the event regions does not match what is rendererd.

* LayoutTests/interaction-region/event-region-overflow-expected.txt: Added.
* LayoutTests/interaction-region/event-region-overflow.html: Added.
Add a test case.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::collectEventRegionForFragments):
Clip to the dirty rect if flag is set, similar to what RenderLayer does elsewhere.

* LayoutTests/fast/scrolling/mac/negative-z-index-overflow-scroll-expected.txt:
* LayoutTests/editing/editable-region/fixed-and-absolute-contenteditable-scrolled-expected.txt:
* LayoutTests/editing/editable-region/relative-inside-fixed-contenteditable-scrolled-expected.txt:
* LayoutTests/editing/editable-region/transformed-scrolled-on-top-of-fixed-contenteditables-expected.txt:
Update event region expectation as it matches the containing bounds, which seems
like a progression.

* LayoutTests/tiled-drawing/scrolling/non-fast-region/wheel-handler-in-columns-expected.txt:
Previous event regions were incorrect based on visual analysis. Update expectations.

Canonical link: https://commits.webkit.org/253580@main
